### PR TITLE
add context support for custom device

### PR DESCRIPTION
**Pull request by heidongxianhua**
Fixes #ISSUE_NUMBER
as the title, add context support for custom device and testcase.
And in the future, we may want to refactor these hooks for different device to unify the APIs, would you agree my 
idea？ @albanD

### CODE REVIEW

**Comment by heidongxianhua** [comment]
@albanD  sorry to bother, could you have a look?

**Review by albanD — COMMENTED** [review]
Hey!
Adding this sounds good but I'm not sure about the implementation. You're adding an indirection where there is already one!

I think you don't need the generic DeviceHooksInterface.h at all. You can have a PrivateUse1HooksInterface.h there that defines the interface, the args and the `REGISTER_PRIVATEUSE1_HOOKS`.
Then in your cpp code you just need to implement these hooks and call that macro.

I think the last one we added is the one for MPS, I would re-use the same logic (where the whole part that lives in the mps/ folder would live out-of-tree here).

As a side note, big thanks for adding the proper testing via cpp extension, that makes the PR a lot easier to read! :tada:

**Comment by heidongxianhua** [comment]
> Hey! Adding this sounds good but I'm not sure about the implementation. You're adding an indirection where there is already one!
> 
> I think you don't need the generic DeviceHooksInterface.h at all. You can have a PrivateUse1HooksInterface.h there that defines the interface, the args and the `REGISTER_PRIVATEUSE1_HOOKS`. Then in your cpp code you just need to implement these hooks and call that macro.
> 
> I think the last one we added is the one for MPS, I would re-use the same logic (where the whole part that lives in the mps/ folder would live out-of-tree here).
> 
> As a side note, big thanks for adding the proper testing via cpp extension, that makes the PR a lot easier to read! 🎉

yeah, I have made changes. @albanD

**Comment by heidongxianhua** [comment]
sorry to bother you,could you have a look again? @albanD

File: aten/src/ATen/detail/PrivateUse1HooksInterface.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#include <ATen/core/Generator.h>
+#include <c10/core/Device.h>
+#include <c10/util/Exception.h>
+namespace at {
+
+struct TORCH_API PrivateUse1HooksInterface {
+  virtual ~PrivateUse1HooksInterface() = default;
+  virtual const at::Generator& getDefaultGenerator(c10::DeviceIndex device_index) {
+    C10_THROW_ERROR(NotImplementedError, "must be inherited to implement.");
```
Comment (albanD) [inline aten/src/ATen/detail/PrivateUse1HooksInterface.h:11]: You can use TORCH_CHECK or TORCH_CHECK_NOT_IMPLEMENTED (if you want a NotImplementedError type on the Python side) here.
Also the error should be that there is no PrivateUse1 device registered (similar to https://github.com/pytorch/pytorch/blob/02cd971e958f4e2f9073e5021501e88cd0460731/aten/src/ATen/detail/CUDAHooksInterface.h#L73)?

File: aten/src/ATen/detail/PrivateUse1HooksInterface.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  virtual at::Device getDeviceFromPtr(void* data) const {
+    C10_THROW_ERROR(NotImplementedError, "must be inherited to implement.");
+  }
+};
+
+struct TORCH_API PrivateUse1HooksArgs {};
+
+TORCH_API void SetPrivateUse1HooksInterface(at::DeviceType t, at::PrivateUse1HooksInterface* hook_);
+
+TORCH_API at::PrivateUse1HooksInterface* GetPrivateUse1HooksInterface(const at::DeviceType& t);
```
Comment (albanD) [inline aten/src/ATen/detail/PrivateUse1HooksInterface.h:23]: Why do you take a devicetype as input here if it is asserted to be pu1? 
Also for consistency with others, should this be called GetPrivateUse1Hooks() ?

**Review by albanD — COMMENTED** [review]
nit on the API but SGTM otherwise!

File: aten/src/ATen/detail/PrivateUse1HooksInterface.h (hunk 2 above)
Comment (heidongxianhua) [reply aten/src/ATen/detail/PrivateUse1HooksInterface.h]: @albanD yeah, I have not noticed that, got it.

Comment (heidongxianhua) [reply aten/src/ATen/detail/PrivateUse1HooksInterface.h]: I have fixed it, and maybe you could have a look again. @albanD

**Review by albanD — APPROVED** [review]
Sounds good! Thanks

**Pull request closed without merging**